### PR TITLE
Add nlargest/nsmallest for DataFrame

### LINKED ISSUE #7359: ENH/API: add nlargest/nsmallest to DataFrame
xref: #7356

API like:

```python
df.nlargest('column_name', 3)
```

impl like:

```python
def nlargest(self, col, n):
    return self.loc[self[col].nlargest(n).index]
```

cc @hayd

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #7359

### CODE REVIEW

**Comment by jreback** [comment]
is there an open issue in pandas for this?

can you show the usage in the top section (and maybe an example in doc-string)?
maybe also add to the docs (I think its in basics.rst)

**Comment by jorisvandenbossche** [comment]
Issue is #7359

**Comment by cpcloud** [comment]
@jreback added a whatsnew, doc example and docstring example

**Comment by jreback** [comment]
should this have nan for values that are not the nlargest/smallest? (eg use where rather than loc)

**Comment by cpcloud** [comment]
i'm not sure what you mean. can you give an example?

**Comment by jreback** [comment]
Was thinking of the utility of this, but your impl is better / more intuitive (as it give the correct ordering, not just the subset).

so nvm.

```
In [19]: df.select_dtypes(include=[np.number]).apply(lambda x: Series.nlargest(x,3))
Out[19]: 
    a   c
1  10   2
2   8 NaN
3  11   3
4 NaN   4

In [20]: df.nlargest(3,'a')
Out[20]: 
    a  b   c
3  11  c   3
1  10  b   2
2   8  d NaN

In [21]: df.nlargest(3,'a').reindex_like(df)
Out[21]: 
    a    b   c
0 NaN  NaN NaN
1  10    b   2
2   8    d NaN
3  11    c   3
4 NaN  NaN NaN
```

File: doc/source/basics.rst [hunk 1]
```diff
@@ -1364,6 +1364,20 @@ faster than sorting the entire Series and calling ``head(n)`` on the result.
...
+.. versionadded:: 0.17.0
+
+``DataFrame`` also has the ``nlargest`` and ``nsmallest`` methods.
+
+.. ipython:: python
+
+   df = DataFrame({'a': [1, 10, 8, 11, -1],
+                   'b': list('abdce'),
+                   'c': [1.0, 2.0, np.nan, 3.0, 4.0]})
+   df.nlargest(3, 'a')
```
Comment (jreback) [inline doc/source/basics.rst]: maybe make the example have some dupes for 'a' (so that ['a','c'] is not obvious)

Comment (cpcloud) [inline doc/source/basics.rst]: done

File: pandas/tests/test_frame.py [hunk 2]
```diff
@@ -14509,6 +14509,40 @@ def _constructor(self):
         # GH9776
         self.assertEqual(df.iloc[0:1, :].testattr, 'XXX')
 
+    def test_nlargest(self):
+        from string import ascii_lowercase
```
Comment (jreback) [inline pandas/tests/test_frame.py]: add the issue number to these as a comment (or put it before once is ok too)

**Comment by jreback** [comment]
give a nice squash as well

**Comment by cpcloud** [comment]
lemme dust off the ol' rebase

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3068,6 +3068,71 @@ def sortlevel(self, level=0, axis=0, ascending=True,
...
+        columns = list(columns)
+        ser = getattr(self[columns[0]], method)(n, take_last=take_last)
+        ascending = dict(nlargest=False, nsmallest=True)[method]
+        return self.loc[ser.index].sort(columns, ascending=ascending,
+                                        kind='mergesort')
+
+    def nlargest(self, n, columns, take_last=False):
+        """Get the rows of a DataFrame sorted by the `n` largest
+        values of `columns`.
+
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Can you add a `.. versionadded:: 0.17.0` here (and for nsmallest as well)?

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3068,6 +3068,71 @@ def sortlevel(self, level=0, axis=0, ascending=True,
...
+        Parameters
+        ----------
+        columns : list or str
+            Column name or names to order by
+        n : int
+            Number of items to retrieve
+
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: the `take_last` argument is not yet included in the Parameters section

**Comment by mrocklin** [comment]
What is the status on this?

**Comment by jreback** [comment]
@cpcloud needs to update for comments
otherwise lgtm

**Comment by cpcloud** [comment]
@jreback @jorisvandenbossche @mrocklin ok 2 merge on passing?

**Comment by jreback** [comment]
lgtm. just add to API.rst as well

**Comment by jorisvandenbossche** [comment]
@cpcloud you have some doc build failures in basics.rst (you can see it in the logs of the third travis build). Needs a `pd.` before Series and DataFrame

For the rest, lgtm!

**Comment by cpcloud** [comment]
@jorisvandenbossche thanks!

**Comment by jreback** [comment]
@cpcloud thanks!

**Pull request merged**